xmas-lights-flask/app.py

```diff
@@ -0,0 +1,225 @@
+import app_html
+
+
+max_leds = 40
+# led_strip = neopixel.NeoPixel(max_leds, 0, 22, 'GRB')  # TODO
+
+# Global scope for script interpretation
+GLOBAL_SCOPE = {
+    'math': math,
+    'random': random,
+    '__builtins__': {
+        'abs': abs,
+        'all': all,
+        'any': any,
+        'bin': bin,
+        'bool': bool,
+        'bytearray': bytearray,
+        'bytes': bytes,
+        'callable': callable,
+        'chr': chr,
+        'complex': complex,
+        'delattr': delattr,
+        'dict': dict,
+        'dir': dir,
+        'divmod': divmod,
+        'enumerate': enumerate,
+        'filter': filter,
+        'float': float,
+        'frozenset': frozenset,
+        'getattr': getattr,
+        'globals': globals,
+        'hasattr': hasattr,
+        'hash': hash,
+        'help': help,
+        'hex': hex,
+        'id': id,
+        'int': int,
+        'isinstance': isinstance,
+        'issubclass': issubclass,
+        'iter': iter,
+        'len': len,
+        'list': list,
+        'locals': locals,
+        'map': map,
+        'max': max,
+        'memoryview': memoryview,
+        'min': min,
+        'next': next,
+        'object': object,
+        'oct': oct,
+        'ord': ord,
+        'pow': pow,
+        'print': print,
+        'property': property,
+        'range': range,
+        'repr': repr,
+        'reversed': reversed,
+        'round': round,
+        'set': set,
+        'setattr': setattr,
+        'slice': slice,
+        'sorted': sorted,
+        'str': str,
+        'sum': sum,
+        'super': super,
+        'tuple': tuple,
+        'type': type,
+        'zip': zip,
+    }
+}
+
+
+reset = True
+
+patterns = {}
+
+try:
+    pattern_filename = Path(__file__).parent / "patterns.json"
+    with open(pattern_filename) as file:
+        patterns = json.loads(file.read())
+except (OSError, ValueError):
+    pass
+
+
+
+def led_thread():
+    global reset
+    global patterns
+    global max_leds
+    current_pattern = None
+    script = None
+    time_seconds = 0
+    timestep = 0.1
+
+    # led = ... TODO: Actually set LEDs
+
+    while True:
+        if reset:
+            reset = False
+            time_seconds = 0
+            current_pattern = None
+            script = None
+            for pattern in patterns.values():
+                if pattern['active']:
+                    current_pattern = pattern
+                    current_pattern['error'] = None
+                    try:
+                        script = compile(current_pattern['script'], current_pattern['name'], 'exec')
+                    except SyntaxError as e:
+                        pattern['active'] = False
+                        pattern['error'] = str(e) # TODO: Get e.line and highlight in GUI
+                        current_pattern = None
+                    local_scope = {'max_leds': max_leds}
+                    break
+        else:
+            time.sleep(timestep)
+            time_seconds += timestep
```

Comment:
TODO, im not actually sure if this is sufficient, given sleep can be for longer than timestep